### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by ApodemusSylvaticus** [openglobus/openglobus-react#8]
Hello, I've encountered a memory leak problem. Over time, the memory runs out and the browser crashes. Only the azimuth changes (approximately every half second). Could you please suggest what exactly the problem might be? And question number 2: is there a way to calculate calculateEndPointByAzimuth faster? Or is there some way to display the direction from the Billboard without rotating it? Thank you in advance for your answer, and I really appreciate your excellent product.
```
import React, { useEffect, useState } from 'react';
import { Billboard, Entity, Geometry, Vector } from '@openglobus/openglobus-react';
import { useMarkerStore } from '../../store/markers.ts';
import { useGlobusStore } from '../../store/globus.ts';
import { useModalStore } from '../../store/modals.ts';
import { LonLat } from '@openglobus/og';

export const calculateEndPointByAzimuth = (startPoint: LonLat, azimuth: number, length: number): LonLat => {
  const R = 6371000;
  const d = length / R;
  const lat1 = (startPoint.lat * Math.PI) / 180;
  const lon1 = (startPoint.lon * Math.PI) / 180;
  const bearing = (azimuth * Math.PI) / 180;

  let lat2 = Math.asin(Math.sin(lat1) * Math.cos(d) + Math.cos(lat1) * Math.sin(d) * Math.cos(bearing));
  let lon2 = lon1 + Math.atan2(Math.sin(bearing) * Math.sin(d) * Math.cos(lat1), Math.cos(d) - Math.sin(lat1) * Math.sin(lat2));

  lat2 = (lat2 * 180) / Math.PI;
  lon2 = (lon2 * 180) / Math.PI;

  return new LonLat(lon2, lat2);
};

export const LineMarker = React.memo(() => {
  const selfMarker = useMarkerStore((state) => state.selfMarker);
  const [end, setEnd] = useState<LonLat>();
  const [rerenderKey, setRerenderKey] = useState(0);
  const activeLayer = useGlobusStore((state) => state.activeLayer);
  const openMarkerInfoModal = useModalStore((state) => state.openMarkerInfoModal);
  const azimuth = useGlobusStore((state) => state.azimuth);

  useEffect(() => {
    const intermediatePoint = calculateEndPointByAzimuth(new LonLat(selfMarker.coords.lon, selfMarker.coords.lat), azimuth, 30000);
    console.log('intermediatePoint', intermediatePoint);
    setEnd(intermediatePoint);
    setRerenderKey((prev) => prev + 1);
  }, [selfMarker, azimuth]);

  useEffect(() => {
    setRerenderKey((prev) => prev + 1);
  }, [activeLayer]);

  return (
    <>
      <Vector
        name={'SelfMarker'}
        clampToGround={true}
        onLclick={() => {
          openMarkerInfoModal({
            key: 'selfMarker',
            coords: selfMarker.coords,
            timeStamp: -1,
            notes: selfMarker.notes,
            tags: [],
            target: { value: 'default_self', src: '', type: 'self' },
          });
        }}
        onTouchEnd={() => {
          openMarkerInfoModal({
            key: 'selfMarker',
            coords: selfMarker.coords,
            timeStamp: -1,
            notes: selfMarker.notes,
            tags: [],
            target: { value: 'default_self', src: '', type: 'self' },
          });
        }}
      >
        <Entity
          name="Self Marker"
          geometry={{ type: 'POINT', coordinates: [selfMarker.coords.lon, selfMarker.coords.lat, selfMarker.coords.alt] }}
          lon={selfMarker.coords.lon}
          lat={selfMarker.coords.lat}
          alt={selfMarker.coords.alt}
          properties={{ color: '#9d2626' }}
        >
          <Billboard size={[30, 30]} src={'./point.svg'} color={'#9d2626'} />
        </Entity>
      </Vector>

      {end && (
        <Vector name={Line_self_mark} key={Line_self_mark_${rerenderKey}}>
          <Entity name={Line_self_mark_entity} lon={selfMarker.coords.lon} lat={selfMarker.coords.lat} alt={0}>
            <Geometry
              lineColor={'rgba(59, 6, 6, 1)'}
              lineWidth={5}
              type={'LINESTRING'}
              coordinates={[
                [selfMarker.coords.lon, selfMarker.coords.lat, 0],
                [end.lon, end.lat, 0],
              ]}
            />
          </Entity>
        </Vector>
      )}
    </>
  );
});
```

**Comment by Zemledelec** [comment]
Hi @ApodemusSylvaticus! 

1) Checking

2) I am afraid not, not available out of the box, fast math is available only inside shaders. As an alternative try GeoObject. It has pitch, yaw, roll,  and scaling parameters, making it visible from any distance the same size. Also, GeoObject orienting is much faster than calculateEndPointByAzimuth.

**Comment by ApodemusSylvaticus** [comment]
@Zemledelec 
This variant (I changed the location of the keys) works as needed.  But I'm not sure if it should work like this
```
 <Vector name={`Line_self_mark`}>
          <Entity key={vectorKey} name={`Line_self_mark_entity`} lon={selfMarker.coords.lon} lat={selfMarker.coords.lat} alt={0}>
            <Geometry
              key={vectorKey}
              lineColor={'rgba(59, 6, 6, 1)'}
              lineWidth={5}
              type={'LINESTRING'}
              coordinates={[
                [selfMarker.coords.lon, selfMarker.coords.lat, 0],
                [end.lon, end.lat, 0],
              ]}
            />
          </Entity>
        </Vector>
```

**Comment by Zemledelec** [comment]
Good, nevertheless, I'll check the engine. @pavletto we need your help with react explanation. 
FYI @ApodemusSylvaticus Geometry (LINESTRING, POLYGON etc.) suits bad for changing its values.  Why you don't use polyline instead?

**Comment by pavletto** [comment]
Hi @ApodemusSylvaticus,
1. I need more time and information to research the problem. I see it in your project. Could you push changes with an azimuth change every 0.5 seconds in a separate brunch? It would be helpful. I want to point out that you're creating marks in a strange way—it would be better if you didn't create a new Vector layer for every mark. Think of `Vector` as a container for your labels. `Vector` can contain a large number of labels.
2. You can use embeded openglobus function [direct](https://openglobus.github.io/docs/Ellipsoid.html#direct). But I agree with @Zemledelec it seems like you can do it easier using other primitives. try to describe what you want to get and we will help you

PS I appreciate your feedback and warm words about the product.

**Comment by ApodemusSylvaticus** [comment]
1.  Here's the branch with the update simulation: https://github.com/ApodemusSylvaticus/sych/tree/bag_preview. The problem is located here: src/components/markers/myPositionMarker, and the update is here: src/hooks/useBroadcast.ts. Also, thank you for your comment, I will review my approach and code
2. .My task was to create a dotted line that shows the direction of myPositionMarker, ideally scaling depending on the height. Due to time constraints, I decided to simplify it as much as possible and encountered this bug. Now I understand that it's better to use GeoObject.

**Comment by pavletto** [comment]
Thanks for your branch - we solved the problem [(v0.1.4).](https://github.com/openglobus/openglobus-react/releases/tag/v0.1.4) It was problem in core during removal process of `Vector` layer. So the fact than you use this layer not as container for entities helped us to found it. Pls let us know if you realise than you need some base component and you dont have it in this package
